src/analysis/natural_parameterization.lean

```diff
@@ -0,0 +1,229 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Rémi Bottinelli
+-/
+import measure_theory.measure.lebesgue
+import analysis.calculus.monotone
+import data.set.function
+import analysis.bounded_variation
+import tactic.swap_var
+/-!
+# Natural Parameterization
+
+This file defines the notion of linear and parameterization for a function `f : ℝ → E` with
+pseudo-emetric structure on `E` with respect to a set `s : set ℝ` and "speed" `l : ℝ`, and shows
+that if `f` has locally bounded variation on `s`, it can be obtained (up to distance zero, on `s`),
+as a composite `φ ∘ (variation_on_from_to f s a)`, where `φ` is linearly parameterized and `a ∈ s`.
+
+## Main definitions
+
+* `is_linearly_parameterized_on_by f s l`, stating that the speed of `f` on `s` is `l`.
+* `is_linearly_parameterized_on f s`, stating that the speed of `f` on `s` is `1`.
+* `natural_parameterization f s a : ℝ → E`, the natural reparameterization of `f`
+
+## Main statements
+
+* `unique_natural_parameterization_on_Icc_zero` proves that if `f` and `f ∘ φ` are both naturally
+  parameterized on closed intervals starting at `0`, then `φ` must be the identity on
+  those intervals.
+* `natural_parameterization_edist_zero` proves that if `f` has locally bounded variation, then
+  precomposing `natural_parameterization f s a` with `variation_on_from_to f s a` yields a function
+  at distance zero from `f` on `s`.
+* `natural_parameterization_is_naturall_parameterized` proves that if `f` has locally bounded
+  variation, then `natural_parameterization f s a` is indeed naturally parameterized on `s`.
+
+## Tags
+
+arc-length, parameterization
+-/
+
+open_locale big_operators nnreal ennreal
+open set measure_theory classical
+
+variables {α : Type*} [linear_order α]{E : Type*} [pseudo_emetric_space E]
+
+/--
+`f` is linearly parameterized on `s` by `l` if the variation of `f` on `s ∩ Icc x y` is equal to
+`l * (y - x)` for any `x y` in `s`.
+This only really makes sense when `l≥0`.
+-/
+def is_linearly_parameterized_on_by (f : ℝ → E) (s : set ℝ) (l : ℝ) :=
+∀ ⦃x⦄ (hx : x ∈ s) ⦃y⦄ (hy : y ∈ s), evariation_on f (s ∩ Icc x y) = ennreal.of_real (l * (y - x))
+
+lemma is_linearly_parameterized_on_by.has_locally_bounded_variation_on
+  {f : ℝ → E} {s : set ℝ} {l : ℝ} (h : is_linearly_parameterized_on_by f s l) :
+  has_locally_bounded_variation_on f s := λ x y hx hy,
+by simp only [has_bounded_variation_on, h hx hy, ne.def, ennreal.of_real_ne_top, not_false_iff]
+
+lemma is_linearly_parameterized_on_by_of_subsingleton
+  (f : ℝ → E) {s : set ℝ} (hs : s.subsingleton) (l : ℝ) : is_linearly_parameterized_on_by f s l :=
+begin
+  rintro x hx y hy, cases hs hx hy,
+  rw evariation_on.subsingleton f (λ y hy z hz, hs hy.1 hz.1 : (s ∩ Icc x x).subsingleton),
+  simp only [sub_self, mul_zero, ennreal.of_real_zero],
+end
+
+-- Could do without `hl` but it means we have to check whether `s` is subsingleton
+-- and if not deduce that `hl` necessarily holds.
```

Comment:
If it's possible, you should definitely do it. Another option is to change the definition to take `l` as an `nnreal` there.

Reply:
I went with making `l` nnreal.